scripts/build.py

```diff
@@ -0,0 +1,243 @@
+#!/usr/bin/env python
+
+# Copyright 2020 Google
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Assuming pre-requisites are in place (please run `scripts/install_preqreqs.py` once on the system), 
+this script will build the firebase cpp sdk
+
+It does the following,
+- install platform specific cpp dependencies via vcpkg
+- cmake configure
+- cmake build
+
+Usage examples:
+# Build all targets with x64 architecture, Release config and 8 cpu cores fo
+python build.py --arch x64 --config Release -j 8
+
+# Build all targets with default options and also build unit tests
+python build.py --build_tests --arch x64
+
+# Build only firebase_app and firebase_auth with default options and also build unit tests
+python build.py --build_tests --arch x64 --targets firebase_app firebase_auth
+
+"""
+
+import subprocess
+import os
+import sys
+import argparse
+import platform
+
+
+def run_command(cmd, cwd=None, as_root_user=False, collect_output=False):
+ """Run a command.
+
+ Args:
+  cmd (list(str)): Command to run as a list object.
+          Eg: ['ls', '-l'].
+  cwd (str): Directory to execute the command from.
+  as_root_user (bool): Run command as root user with admin priveleges (supported on mac and linux).
+  collect_output (bool): Get output from the command. The string returned from this function
+       call will have the output from command execution.
+ Raises:
+  (subprocess.CalledProcessError): If command errored out.
+ Returns:
+  (str|None): If collect_output is provided, a string is returned
+     None if output is not requested.
+ """
+ if as_root_user and (is_mac_os() or is_linux_os()):
+  cmd.insert(0, 'sudo')
+
+ cmd_string = ' '.join(cmd)
+ print('Running cmd: {0}\n'.format(cmd_string))
+ if collect_output:
+  output = subprocess.check_output(cmd, cwd=cwd)
+  return output.decode('utf-8').strip()
+ else:
+  subprocess.call(cmd, cwd=cwd)
```

Comment:
Your comment mentions using "Python 3" goodness but is still written in a Python 2 friendly way.

If we're fine with requiring Python 3, there is a unified subprocess method called subprocess.run, which has args to reproduce the various behaviours of old subprocess methods. check, capture_output, etc. You can also specify the text or encoding arg instead of decoding the output explicitly.

https://docs.python.org/3/library/subprocess.html#using-the-subprocess-module

Reply:
Yeah let's use that, the GHA runners have Python 3, and I'm okay with listing it as a dependency if users want to run the scripts manually.

Reply:
Good to know. Yeah, aside from this, I used some other Python3 "only" constructs,
- downloading urls
- unpacking files
- os.cpu_count instead of multiprocess.cpu_count

So yeah, Python3 is definitely required for these scripts.